### DISCUSSION TITLE
How to implement a recommender system for a tools ecommerce store with limited user data?

### DISCUSSION THREAD

**Discussion by filiptrivan** [recommenders-team/recommenders#2162] [Q&A] [answered]
I have an ecommerce site in the tools industry with about 60k products, but only a small percentage have reviews. We get around 5k visits per day, mostly from users who are not logged in, and we save the history of reviewed products in local storage. Given these constraints, what are the best approaches to implement a recommender system focused on product recommendations? Should I consider changing my infrastructure to collect more user data, and if so, how?

**Comment by daviddavo** [comment, marked as answer]
What you describe is very common, if you plot the distribution of popularity, you get a very long tail.

![image](https://github.com/user-attachments/assets/e59d9f50-7901-4358-88fc-f76d4e47b2a2)

There are two main approaches to recommend items that I will present with two over-simplified examples:
- Collaborative Filtering: "User A buys X and Y, user B buys X, so let's recommend Y to B"
- Content-based: "User B buys items from the _screwdrivers_ section, let's recommend items in that section"

If you have lot of data on the items, you should go on the content-based approach. You can also combine both approaches in hybrid systems or use two recommenders and display them one on top of another in the web ui.

In your case, where you don't have a lot of data from each user and it is stored in local storage, you can check out session-based recommender systems, which focus on recommending the "next item".

Personally, I'd recommend you to try a couple models with the offline dataset, to check what works better.

About the collecting more data part, you can use not only the history of reviewed products (explicit feedback) but the history of clicked/wishlisted/bought products (implicit feedback) to generate the recommendations.

**Reply by filiptrivan** [reply]
Thanks, I was a bit slow with the implementation due to other obligations, but I think we’re close to the finish - https://github.com/filiptrivan/pa-recommender. I used Ben Fred’s implicit feedback library with the ALS algorithm - https://github.com/benfred/implicit.
